spec/models/manifest_uploader_spec.rb

```diff
@@ -61,4 +65,103 @@
+    let!(:manifest) { build(:csv_manifest, locations: [ordered_location], number_of_labwares: 3, labware_prefix: labware_prefix, positions: ["", 1.5, "string"]).generate_csv }
+
+    attr_reader :data
+
+    before(:each) do
+      manifest_uploader.file = manifest
+    end
+
+    it 'will not be valid' do
+      expect(manifest_uploader).to_not be_valid
+    end
+
+    it 'will show an error' do
+      manifest_uploader.run
+      expect(manifest_uploader.errors.full_messages).to contain_exactly("Line 2: invalid entry for position. Please specify a positive integer.",
+                                                                        "Line 3: invalid entry for position. Please specify a positive integer.",
+                                                                        "Line 4: invalid entry for position. Please specify a positive integer.")
+    end
+  end
+
+  context 'when a position does not exist' do
+    let!(:manifest) { build(:csv_manifest, locations: [ordered_location], number_of_labwares: 1, labware_prefix: labware_prefix, positions: [invalid_position]).generate_csv }
+
+    attr_reader :data
+
+    before(:each) do
+      manifest_uploader.file = manifest
+    end
+
+    it 'will not be valid' do
+      expect(manifest_uploader).to_not be_valid
+    end
+
+    it 'will show an error' do
+      manifest_uploader.run
+      expect(manifest_uploader.errors.full_messages).to include("Line 2: target position #{invalid_position} for location with barcode #{ordered_location.barcode} does not exist")
+    end
+  end
+
+  context 'when there are duplicate positions entered' do
+    let!(:manifest) { build(:csv_manifest, locations: [ordered_location], number_of_labwares: 2, labware_prefix: labware_prefix, positions: [1, 1]).generate_csv }
+
+    attr_reader :data
+
+    before(:each) do
+      manifest_uploader.file = manifest
+    end
+
+    it 'will not be valid' do
+      expect(manifest_uploader).to_not be_valid
+    end
+
+    it 'will show an error' do
+      manifest_uploader.run
+      expect(manifest_uploader.errors.full_messages).to include("Lines 2,3: duplicate target positions")
+    end
+  end
+
+  context 'when there is a position which is already occupied' do
+    let!(:existing_labware) { create(:labware, location_id: ordered_location.id, coordinate_id: Coordinate.find_by(location_id: ordered_location.id, position: 1).id) }
+    let!(:manifest) { build(:csv_manifest, locations: [ordered_location], number_of_labwares: 1, labware_prefix: labware_prefix, positions: [1]).generate_csv }
+
+    attr_reader :data
+
+    before(:each) do
+      manifest_uploader.file = manifest
+    end
+
+    it 'will not be valid' do
+      expect(manifest_uploader).to_not be_valid
+    end
+
+    it 'will show an error' do
+      manifest_uploader.run
+      expect(manifest_uploader.errors.full_messages).to include("Line 2: target position 1 for labware with barcode RNA000001 is already occupied")
+    end
+  end
```

Comment:
Some nice validation and error messages 👍

Reply:
Thanks 😁